### DISCUSSION TITLE
BPF/eBPF vs async profiler

### DISCUSSION THREAD

**Discussion by kibor** [async-profiler/async-profiler#529] [Ideas]
Hey folks,

Sorry for the dumb question, but I really can't figure that out on my own.

There is a technology that allows to profile CPU on Linux (Windows is in progress) without any dependency to the programming language. Here is the example - https://blog.px.dev/cpu-profiling-2/

Is this technology better than the one that is used for async profiler?
Is it actually worse in some way? For now I only found system restriction (Linux only, new kernel) and I can't find examples where it is used to profile memory leaks and thread contention.
Does async profiler actually uses BPF or maybe there are plans to somehow make use of it?

**Comment by kibor** [comment]
No one likes me :(

**Comment by apangin** [comment]
Not at all - this is actually a good question, but I it takes time to give a comprehensive answer - will do it some time soon.
In short, these are technologies for different use cases; eBPF is neither better nor worse, there are both advantages and drawbacks.

**Reply by krzysztofslusarski** [reply]
I believe that the major drawbacks of eBPF in "classic" Java CPU profiling is that, to collect java stacktrace using eBPF you still need the ```-XX:+PreserveFramePointer``` that causes additional overhead. And ```perf-map-agent``` ...

You can do magic using eBPF, there is a book by Brendan Gregg explaining it. But still it is a technology that observes processes from OS point of view. Async-profiler is a tool dedicated to JVM. It has features that are tailored especially for that ecosystem, for example:
- heap allocation profiling
- wall clock profiling with java stacktrace
- java method profiling

I leave a pleasure of "more comprehensive answer" to @apangin :)

**Comment by apangin** [comment]
@krzysztofslusarski's answer is quite correct; I'll just extend it a bit.

eBPF is a technology; async-profiler is a product - it's not quite correct to compare them.

Even though async-profiler currently does not use eBPF, it still provides many advanced profiling capabilities: kernel stack traces, accurate unbiased sampling, hardware performance counters, kprobes, uprobes, etc. It's absolutely possible to create a low overhead profiler for Java without eBPF; however, the profiler may benefit from eBPF in some cases. There is an ongoing work of bringing the optional eBPF support to async-profiler - you might have noticed `bpfclient` branch in the repo.

So what are the advantages and drawbacks of using eBPF?

### Advantages

The technology allows to execute custom code snippets in the kernel context. This brings us to the main advantage - you can do profiling directly in kernel without an overhead of switching to user mode and passing data between kernel and userland. However, these snippets are very restricted in what they can do (see drawbacks below).

The overhead of async-profiler is already low enough to use it in production, even for continuous profiling, as long as you run one-two-three-five Java applications on a server. Now imagine a cloud server that runs hundreds of Java applications at a time. If you start profiling all of them simultaneously, system CPU usage may get too high due to thousands of perf_event objects competing for the limited number of CPUs. Here is where eBPF may help - a single agent can do system-wide profiling while taking care of each process individually. So the use case for eBPF in async-profiler will be to continuously profile multiple Java applications on the same machine in production.

### Drawbacks

1. Requires a rather new Linux kernel. E.g. the feature I described above will work only on 5.10+ kernels.
2. Loading a eBPF program typically requires root privileges (more precisely, CAP_BPF).
3. eBPF can't have loops, they are limited in size, they are not allowed to access arbitrary userland structures. In short, they can do only very simple things. But walking through a Java stack is not that simple.
4. As a consequence, eBPF programs are not Java friendly - in general, they can't get Java stack traces.
5. To work around (4), you may start JVM with `-XX:+PreserveFramePointer` option, which forces JVM to maintain the stack structure that eBPF can walk. However, this option may degrade application performance by several percent.
6. Even with the above option, you can see only JIT-compiled frames. eBPF can hardly obtain the names of interpreted Java methods.
7. The maximum stack depth is limited to 127 by default (but it's tunable via sysctl config).
8. eBPF does not know about Java specifics, and because of the limitations, you can't really "teach" it to do allocation profiling, lock profiling, etc. It can't even get full Java thread names.

**Reply by kibor** [reply]
Thanks!
Love that kind of detailed answers :)

Please let me know if I can help you with the BPF part, I am also interested in that.

**Reply by Loveen01** [reply]
Lovely answer thank you

**Comment by toktarev** [comment]
I let  myself to explain other people (like me) what is BPF:

The Berkeley Packet Filter (BPF) is a technology used in certain computer operating systems for programs that need to, among other things, analyze network traffic.… Since version 3.18, the Linux kernel includes an extended BPF virtual machine with ten 64-bit registers, termed extended BPF (eBPF). It can be used for non-networking purposes, such as for attaching eBPF programs to various tracepoints. Since kernel version 3.19, eBPF filters can be attached to sockets, and, since kernel version 4.1, to traffic control classifiers for the ingress and egress networking data path.

https://en.wikipedia.org/wiki/Berkeley_Packet_Filter
